test/graph_input_tests.cpp

```diff
@@ -0,0 +1,648 @@
+#include <catch2/catch_test_macros.hpp>
+
+#include "test_graph.h"
+
+#include <gdsb/graph_input.h>
+
+#include <sstream>
+#include <string>
+
+using namespace gdsb;
+
+TEST_CASE("read_ulong")
+{
+    unsigned long const a = 40;
+    unsigned long const b = 200;
+    unsigned long const c = 11;
+    unsigned long const d = 29848572984;
+
+    std::stringstream ss;
+    ss << a << " " << b << " " << c << " " << d << "\n";
+    std::string const test_string = ss.str();
+
+    char const* string_source = test_string.c_str();
+    char* string_position = nullptr;
+
+    unsigned long const a_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const b_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const c_read = read_ulong(string_source, &string_position);
+    string_source = string_position;
+    unsigned long const d_read = read_ulong(string_source, &string_position);
+
+    CHECK(a == a_read);
+    CHECK(b == b_read);
+    CHECK(c == c_read);
+    CHECK(d == d_read);
+}
+
+TEST_CASE("read_graph, edge_list")
+{
+    Edges32 edges;
+    auto emplace = [&](Vertex32 u, Vertex32 v) { edges.push_back(Edge32{ u, v }); };
+
+    WeightedEdges32 weighted_edges;
+    auto emplace_weighted = [&](Vertex32 u, Vertex32 v, Weight w) {
+        weighted_edges.push_back(WeightedEdge32{ u, Target32{ v, w } });
+    };
+
+    std::ifstream undirected_unweighted_temporal(graph_path + undirected_unweighted_temporal_reptilia_tortoise);
+    std::ifstream graph_input_weighted_temporal(graph_path + small_weighted_temporal_graph);
+    std::ifstream graph_input_unweighted_directed(graph_path + unweighted_directed_graph_enzymes);
+
+    SECTION("undirected, unweighted")
+    {
+        auto const [vertex_count, edge_count] =
+            read_graph<Vertex32, decltype(emplace), EdgeListUndirectedUnweightedStatic>(undirected_unweighted_temporal,
+                                                                                        std::move(emplace));
+
+        CHECK(104 * 2 == edges.size());
+        CHECK(104 * 2 == edge_count);
+
+        bool edge_16_to_17_exists = [&]()
+        {
+            for (Edge32 const& e : edges)
+            {
+                if (e.source == 16)
+                {
+                    if (e.target == 17)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }();
+
+        CHECK(edge_16_to_17_exists);
+    }
+
+    SECTION("directed, unweighted")
+    {
+        auto const [vertex_count, edge_count] =
+            read_graph<Vertex32, decltype(emplace), EdgeListDirectedUnweightedStatic>(undirected_unweighted_temporal,
+                                                                                      std::move(emplace));
+
+        // directed: thus original edge count 103
```

Comment:
this off-by-one-error occurred at multiple occasions, maybe worth to investigate

Reply:
`104` is actually correct, thanks for the pointer. I removed the comment lines.